Matheus Silva dos Santos - 22052573 <br>
Felipe Fraxe Filho - 22251894

## Atividade Prática 3: Regras de Parada

### Parte 1 (Regra de Chow e Robbins)

**Objetivo**

O objetivo desta parte da atividade é simular uma fila M/M/1 (chegadas e serviços exponenciais) com a regra de parada de Chow e Robbins. Essa regra permite determinar o número de clientes (n) necessário para atingir uma determinada precisão (d) no intervalo de confiança do tempo médio de espera na fila.

In [9]:
# Importações necessárias

import numpy as np
import random
import ADBib

**Parâmetros do script abaixo:**

- lambd = 9: Taxa de chegada de clientes (9 clientes por segundo).
- mu = 10: Taxa de serviço do servidor (10 clientes por segundo).
- n = 10**3: Valor inicial do número de clientes a serem simulados.
- d: Precisão desejada no intervalo de confiança (valores testados: 1, 0.5, 0.1, 0.05).


In [30]:
lambd = 9
mu = 10

n = 10 ** 3 # Valor inicial
for d in [1, 0.5, 0.1, 0.05]:
    H = d + 1
    while H > d:
        wait_times = np.zeros(n, dtype=np.float32)

        for j in range(1, n):
            service_time = random.expovariate(mu) # Tempo de serviço de j-1
            arrival_time = random.expovariate(lambd) # Tempo de chegada de j

            curr_wait_time = wait_times[j-1] - arrival_time + service_time

            if curr_wait_time < 0:
                curr_wait_time = 0
            wait_times[j] = curr_wait_time

        mean = ADBib.arithmetic_mean(wait_times)
        ci = ADBib.confidence_interval(wait_times, mean, 0.95)
        H = ci[1] - mean
        if H > d:
            n += 100

    print(f"\nPara n = {n}:")
    print(f"Para d = {d}")
    print(f"Tempo Médio Estimado de Espera: {mean:.6f}")
    print(f"Intervalo de Confiança de 95%: {ci}")


Para n = 1000:
Para d = 1
Tempo Médio Estimado de Espera: 0.384401
Intervalo de Confiança de 95%: (0.35997772511668286, 0.4088235510831861)

Para n = 1000:
Para d = 0.5
Tempo Médio Estimado de Espera: 0.552739
Intervalo de Confiança de 95%: (0.5232713572959228, 0.5822074936966483)

Para n = 1100:
Para d = 0.1
Tempo Médio Estimado de Espera: 1.034179
Intervalo de Confiança de 95%: (0.969434599332202, 1.098923685688708)

Para n = 1100:
Para d = 0.05
Tempo Médio Estimado de Espera: 0.750704
Intervalo de Confiança de 95%: (0.7129962580596803, 0.7884124695984067)


**Conclusão dos Experimentos**

Como os resultados de cada execução são aleatórios, vamos considerar os resultados abaixo para realizar a análise:

Para n = 1000: <br>
Para d = 1 <br>
Tempo Médio Estimado de Espera: 0.384401 <br>
Intervalo de Confiança de 95%: (0.35997772511668286, 0.4088235510831861)

Para n = 1000: <br>
Para d = 0.5 <br>
Tempo Médio Estimado de Espera: 0. 552739 <br>
Intervalo de Confiança de 95%: (0.5232713572959228, 0.5822074936966483)

Para n = 1100: <br>
Para d = 0.1 <br>
Tempo Médio Estimado de Espera: 1.034179 <br>
Intervalo de Confiança de 95%: (0.969434599332202, 1.098923685688708)

Para n = 1100: <br>
Para d = 0.05 <br>
Tempo Médio Estimado de Espera: 0.750704 <br>
Intervalo de Confiança de 95%: (0.7129962580596803, 0.7884124695984067)

Os resultados acima mostra que, para atingir uma determinada precisão d no intervalo de confiança do tempo médio de espera, é necessário simular um número mínimo de clientes n. Observamos que:

- Quanto menor o valor de d (maior precisão desejada), maior o número de clientes n necessário para que o tamanho do intervalo de confiança seja menor ou igual a d.

- Para um mesmo valor de d, o tempo médio de espera estimado pode variar entre as execuções da simulação, devido à natureza aleatória do processo de chegada e serviço dos clientes.

- O tempo médio de espera estimado tende a se aproximar do valor teórico (0.9 segundos para este sistema M/M/1) à medida que o número de clientes aumenta. No entanto, mesmo com um número relativamente grande de clientes (1100), a estimativa ainda pode estar um pouco distante do valor teórico, especialmente para valores de d menores.

Os resultados da simulação mostram que a precisão desejada tem um impacto significativo no número de clientes necessário e que a convergência para o valor teórico é mais lenta para valores de d menores.

### Parte 2 (Regra do Tamanho Relativo do Intervalo de Confiança)

**Objetivo**

O objetivo desta parte da atividade é simular uma fila M/M/1 (chegadas e serviços exponenciais) com a regra de parada baseada no tamanho relativo do intervalo de confiança. Essa regra permite determinar o número de clientes (n) necessário para atingir uma determinada precisão relativa (γ) na estimativa do tempo médio de espera na fila.



In [31]:
# Importações necessárias

import numpy as np
import random
import matplotlib.pyplot as plt
import ADBib

**Parâmetros do script abaixo:**

- lambd = 9: Taxa de chegada de clientes (9 clientes por segundo).
- mu = 10: Taxa de serviço do servidor (10 clientes por segundo).
- gamma = 0.05: Precisão relativa desejada (5%).
- n = 10**3: Valor inicial do número de clientes a serem simulados.

In [40]:
lambd = 9
mu = 10
gamma = 0.05

n = 10 ** 3 # Valor inicial

# Valores iniciais para garantir a entrada no loop
mean = 1
H = mean
while (H / mean) > gamma:
    wait_times = np.zeros(n, dtype=np.float32)

    for j in range(1, n):
        service_time = random.expovariate(mu) # Tempo de serviço de j-1
        arrival_time = random.expovariate(lambd) # Tempo de chegada de j

        curr_wait_time = wait_times[j-1] - arrival_time + service_time

        if curr_wait_time < 0:
            curr_wait_time = 0
        wait_times[j] = curr_wait_time

    mean = ADBib.arithmetic_mean(wait_times)
    ci = ADBib.confidence_interval(wait_times, mean, 0.95)
    H = ci[1] - mean
    if (H / mean) > gamma:
        n += 100

print(f"\nPara n = {n}:")
print(f"Para H / média = {H / mean:.6f}")
print(f"Tempo Médio Estimado de Espera: {mean:.6f}")
print(f"Intervalo de Confiança de 95%: {ci}")


Para n = 1100:
Para H / média = 0.049927
Tempo Médio Estimado de Espera: 0.646269
Intervalo de Confiança de 95%: (0.6140028537210019, 0.6785355317024394)


**Conclusão dos Experimentos**

Novamente, como os resultados de cada execução são aleatórios, vamos considerar os resultados abaixo para realizar a análise:

Para n = 1100: <br>
Para H / média = 0.049927 <br>
Tempo Médio Estimado de Espera: 0.646269 <br>
Intervalo de Confiança de 95%: (0.6140028537210019, 0.6785355317024394)

A regra de parada do tamanho relativo do intervalo de confiança mostrou-se eficaz em determinar o número de clientes necessário para atingir a precisão relativa desejada de 5%. Com 1100 clientes simulados, a razão H/média ficou abaixo do limite de 0.05, indicando que o intervalo de confiança é suficientemente estreito em relação à média estimada.

O tempo médio de espera estimado (0.6463 segundos) está próximo do valor esperado teórico para uma fila M/M/1 com taxa de chegada λ = 9 e taxa de serviço μ = 10. No entanto, é importante ressaltar que este é apenas um resultado de uma única execução da simulação. Devido à natureza estocástica do processo de chegada e serviço, o tempo médio de espera pode variar em diferentes execuções.

O intervalo de confiança de 95% (0.6140 a 0.6785 segundos) fornece uma medida da incerteza associada à estimativa do tempo médio de espera. Podemos ter 95% de confiança de que o verdadeiro tempo médio de espera da população está dentro desse intervalo.

O resultado da simulação indica que, para o sistema M/M/1 em questão, com as taxas de chegada e serviço especificadas, é necessário simular pelo menos 1100 clientes para obter uma estimativa do tempo médio de espera com uma precisão relativa de 5%. O tempo médio de espera estimado de 0.6463 segundos é uma estimativa pontual, e o intervalo de confiança fornece uma faixa de valores plausíveis para o verdadeiro tempo médio de espera.



### Conclusão Geral

Esta atividade explorou a simulação de uma fila M/M/1 com duas regras de parada distintas: Chow e Robbins (precisão absoluta) e tamanho relativo do intervalo de confiança (precisão relativa). Ambas as regras demonstraram ser eficazes na determinação do número de clientes necessário para obter estimativas do tempo médio de espera com a precisão desejada.

**Parte 1: Regra de Chow e Robbins**

Os resultados mostraram que a precisão desejada (d) influencia diretamente o número de clientes (n) necessário para atingir a condição de parada. Quanto menor o valor de d (maior precisão), maior o número de clientes necessário. A variabilidade nas estimativas do tempo médio de espera foi observada, principalmente para valores menores de n, mas a tendência de convergência para o valor teórico foi evidente à medida que n aumentou.

**Parte 2: Regra do Tamanho Relativo do Intervalo de Confiança**

A regra do tamanho relativo do intervalo de confiança também se mostrou eficaz, convergindo para uma precisão relativa de 5% (γ = 0.05) com 1100 clientes simulados. O tempo médio de espera estimado foi próximo ao valor teórico, e o intervalo de confiança forneceu uma faixa de valores plausíveis para o verdadeiro tempo médio de espera.

**Comparação das Regras de Parada implementadas**

A regra de Chow e Robbins tende a exigir um número maior de clientes para atingir a mesma precisão que a regra do tamanho relativo do intervalo de confiança. Isso ocorre porque a regra de Chow e Robbins busca uma precisão absoluta, enquanto a regra do tamanho relativo busca uma precisão em relação à média estimada.

**Considerações Finais**

A escolha da regra de parada mais adequada depende dos objetivos da simulação e das características do sistema. A regra de Chow e Robbins pode ser preferível quando a precisão absoluta é crítica, enquanto a regra do tamanho relativo pode ser mais vantajosa quando a precisão relativa é suficiente e deseja-se minimizar o número de clientes simulados.

Em ambos os casos, é importante ressaltar que os resultados da simulação são influenciados pela aleatoriedade do processo de chegada e serviço dos clientes.